server/src/main/java/org/eclipse/hono/application/HonoApplication.java

```diff
@@ -0,0 +1,248 @@
+                deployAuthenticationService(), // we only need 1 authentication service
+                deployAuthorizationService(), // we only need 1 authorization service
+                deployRegistrationService()).setHandler(ar -> {
+            if (ar.succeeded()) {
+                deployServer(firstInstance, instanceCount, started);
+            } else {
+                started.fail(ar.cause());
+            }
+        });
+
+        try {
+            if (startupLatch.await(honoConfig.getStartupTimeout(), TimeUnit.SECONDS)) {
+                LOG.info("Hono startup completed successfully");
+            } else {
+                LOG.error("startup timed out after {} seconds, shutting down ...", honoConfig.getStartupTimeout());
+                shutdown();
+            }
+        } catch (InterruptedException e) {
+            LOG.error("startup process has been interrupted, shutting down ...");
+            Thread.currentThread().interrupt();
+            shutdown();
+        }
+    }
+
+    private Future<String> deployRegistrationService() {
+        LOG.info("Starting registration service {}", registrationService);
+        Future<String> result = Future.future();
+        vertx.deployVerticle(registrationService, result.completer());
+        return result;
+    }
+
+    private Future<String> deployAuthenticationService() {
+        LOG.info("Starting authentication service {}", authenticationService);
+        Future<String> result = Future.future();
+        vertx.deployVerticle(authenticationService, result.completer());
+        return result;
+    }
+
+    private Future<String> deployAuthorizationService() {
+        LOG.info("Starting authorization service {}", authorizationService);
+        Future<String> result = Future.future();
+        vertx.deployVerticle(authorizationService, result.completer());
+        return result;
+    }
+
+    private void deployServer(final HonoServer firstInstance, final int instanceCount, Future<Void> startFuture) {
+        @SuppressWarnings("rawtypes")
+        List<Future> results = new ArrayList<>();
+        deployServerInstance(firstInstance, results);
+        for (int i = 1; i < instanceCount; i++) {
+            HonoServer server = serverFactory.getHonoServer();
+            deployServerInstance(server, results);
+        }
+        CompositeFuture.all(results).setHandler(ar -> {
+           if (ar.failed()) {
+              startFuture.fail(ar.cause());
+           } else {
+               startFuture.complete();
+           }
+        });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void deployServerInstance(final HonoServer instance, final List<Future> deployResults) {
+        Future<String> result = Future.future();
+        vertx.deployVerticle(instance, result.completer());
+        deployResults.add(result);
+    }
+
+    /**
+     * Stops the Hono server in a controlled fashion.
+     */
+    @PreDestroy
+    public void shutdown() {
```

Comment:
again, either `final` and provide separate method for subclass customization or remove annotation